### PR TITLE
add quantized ELU activation

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34267 add quantized ELU activation**

Summary:

Adds quantized ELU.

Test Plan:

```
python test/test_quantized.py TestQuantizedOps.test_qelu
```

still need to benchmark, saving that for after the review comments

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20370953](https://our.internmc.facebook.com/intern/diff/D20370953)

### CODE REVIEW

**Comment by t-vi** [comment]
Is there a report somewhere how well this works?
Given that ELU saturates at -1 it would seem that it might run into trouble when -1 isn't represented exactly, much like ReLU needs exact representation of 0.

**Comment by vkuzo** [comment]
> Is there a report somewhere how well this works?
> Given that ELU saturates at -1 it would seem that it might run into trouble when -1 isn't represented exactly, much like ReLU needs exact representation of 0.

yes, good point, thank you!  There is still an open TODO in this PR about the best strategy to take around scale and zero point which I should definitely improve before landing this.  Perhaps we can add optional behavior to choose output scale + zp such as both -1 and 0 are exactly represented given the data type - will look into it.

File: aten/src/ATen/native/quantized/cpu/quantized_ops.h [hunk 1]
```diff
@@ -15,6 +15,10 @@ using qclamp_fn = void (*)(
     Scalar max,
     at::Tensor& /*qy*/);
 using qtanh_fn = void (*)(const at::Tensor& /*qx*/, at::Tensor& /*qy*/);
+using qelu_fn = void(*)(
+    const at::Tensor& /*qx*/,
+    Scalar alpha,
```
Comment (z-a-f) [inline aten/src/ATen/native/quantized/cpu/quantized_ops.h]: nit: remove `alpha` for consistency.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 2]
```diff
@@ -394,6 +394,79 @@ void qtanh_kernel(const Tensor& qx, Tensor& qy) {
...
+        auto dx_vec_vec = value_qx.dequantize(i_scale_vec, i_zero_point_vec,
+                                            i_scale_neg_zp_premul_vec);
+        for (int idx = 0; idx < dx_vec_vec.size(); idx++) {
+
+          // quickly check if any elements are below zero
+          auto cmp_to_zero = dx_vec_vec[idx] > zero_vec;
+
+          // if (!cmp_to_zero.zero_mask()) {
+          //   // do nothing
+          // }
```
Comment (z-a-f) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: nit: cleanup

File: torch/nn/quantized/functional.py [hunk 3]
```diff
@@ -323,6 +323,36 @@ def leaky_relu(input, negative_slope=0.01, inplace=False,
         result = torch._C._nn.leaky_relu(input, negative_slope)
     return result
 
+def elu(input, alpha=1., inplace=False, scale=None, zero_point=None):
+    r"""float(input, alpha_) -> Tensor
```
Comment (z-a-f) [inline torch/nn/quantized/functional.py]: I think this line is not necessary

**Review by z-a-f — APPROVED** [review]
LGTM

**Pull request closed without merging**